File: 998/998.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/998.py

using System;
using System.Collections.Generic;
using System.Numerics;

class Program
{
    static long GcdLong(long a, long b)
    {
        while (b != 0) { long t = b; b = a % b; a = t; }
        return a;
    }

    static long IsqrtLong(long n)
    {
        if (n < 0) throw new ArgumentException("isqrt of negative");
        if (n == 0) return 0;
        long x = (long)Math.Sqrt((double)n);
        while (x * x > n) x--;
        while ((x + 1) * (x + 1) <= n) x++;
        return x;
    }

    static List<long>[] PythagoreanPartners(int limit)
    {
        var partners = new List<long>[limit + 1];
        for (int i = 0; i <= limit; i++) partners[i] = new List<long>();
        long rMax = IsqrtLong(2L * limit) + 3L;
        for (long r = 2; r <= rMax; r++)
        {
            long rr = r * r;
            for (long s = 1; s < r; s++)
            {
                if (((r - s) & 1L) == 0L || GcdLong(r, s) != 1L) continue;
                long a = rr - s * s;
                long b = 2 * r * s;
                long m = a > b ? a : b;
                long x = a > b ? b : a;
                if (m > limit) continue;
                for (long km = m; km <= limit; km += m)
                    partners[(int)km].Add((km / m) * x);
            }
        }
        foreach (var row in partners) row.Sort();
        return partners;
    }

    static bool IsMinimumSquare(long[] sides, long twiceArea, long squareSide)
    {
        long[] ss = sides;
        long m = squareSide;
        long m2 = m * m;
        long dArea = twiceArea;
        bool hasEqualCandidate = false;

        for (int i = 0; i < 3; i++)
        {
            long d = ss[i];
            long e = ss[(i + 1) % 3];
            long f = ss[(i + 2) % 3];
            long den = 2L * d;
            long tNum = d * d + e * e - f * f;
            long dNum = d * den;
            long spanMax = Math.Max(0L, Math.Max(dNum, tNum));
            long spanMin = Math.Min(0L, Math.Min(dNum, tNum));
            long widthNum = spanMax - spanMin;
            long widthCmp = widthNum - m * den;
            long heightCmp = dArea - m * d;
            if (widthCmp < 0L && heightCmp < 0L) return false;
            if (widthCmp <= 0L && heightCmp <= 0L && (widthCmp == 0L || heightCmp == 0L))
                hasEqualCandidate = true;
        }

        for (int i = 0; i < 3; i++)
        {
            long r = ss[i];
            long p = ss[(i + 1) % 3];
            long q = ss[(i + 2) % 3];
            long kNum = p * p + q * q - r * r;
            if (kNum <= 0L) continue;
            long rDenPart = p * p + q * q - 2L * dArea;
            if (rDenPart <= 0L) continue;
            BigInteger num = (BigInteger)kNum * kNum;
            BigInteger den = 4 * (BigInteger)rDenPart;
            BigInteger p2 = (BigInteger)(p * p);
            BigInteger q2 = (BigInteger)(q * q);
            BigInteger dAreaBig = (BigInteger)dArea;
            if (num < dAreaBig * den) continue;
            if (num > p2 * den || num > q2 * den) continue;
            if (p2 * den > 2 * num || q2 * den > 2 * num) continue;
            BigInteger target = (BigInteger)m2 * den;
            if (num < target) return false;
            if (num == target) hasEqualCandidate = true;
        }

        return hasEqualCandidate;
    }

    static long Solve(int limit)
    {
        var partners = PythagoreanPartners(limit);
        var seen = new HashSet<(long, long, long)>();
        long total = 0L;

        for (int mInt = 1; mInt <= limit; mInt++)
        {
            long m = mInt;
            long mm = m * m;
            var row = new List<(long, long)> { (0L, m) };
            foreach (long x in partners[mInt])
                row.Add((x, IsqrtLong(mm + x * x)));
            int rowLen = row.Count;

            for (int i = 0; i < rowLen; i++)
            {
                var (x, hx) = row[i];
                for (int j = i; j < rowLen; j++)
                {
                    var (y, hy) = row[j];
                    long baseVal = x + y;
                    if (baseVal == 0L) continue;
                    if (baseVal > m) break;
                    if (x * y < m * (m - baseVal)) continue;
                    long[] arr = new long[] { baseVal, hx, hy };
                    Array.Sort(arr);
                    var key = (arr[0], arr[1], arr[2]);
                    if (!seen.Contains(key))
                    {
                        seen.Add(key);
                        total += arr[0] + arr[1] + arr[2];
                    }
                }
            }

            for (int i = 0; i < rowLen; i++)
            {
                var (u, hu) = row[i];
                for (int j = i; j < rowLen; j++)
                {
                    var (v, hv) = row[j];
                    long twiceArea = mm - u * v;
                    if (twiceArea <= 0L) continue;
                    long p = m - u;
                    long q = m - v;
                    long third2 = p * p + q * q;
                    long third = IsqrtLong(third2);
                    if (third * third != third2 || third == 0L) continue;
                    long[] arr = new long[] { third, hu, hv };
                    Array.Sort(arr);
                    var key = (arr[0], arr[1], arr[2]);
                    if (seen.Contains(key)) continue;
                    if (IsMinimumSquare(arr, twiceArea, m))
                    {
                        seen.Add(key);
                        total += arr[0] + arr[1] + arr[2];
                    }
                }
            }
        }

        return total;
    }

    static void Main()
    {
        System.Diagnostics.Debug.Assert(Solve(40) == 346L);
        System.Diagnostics.Debug.Assert(Solve(400) == 76402L);
        System.Diagnostics.Debug.Assert(Solve(2000) == 3237036L);
        Console.WriteLine(Solve(1_000_000));
    }
}

File: 998/998.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/998.py

use std::collections::HashSet;

fn gcd(mut a: i64, mut b: i64) -> i64 {
    while b != 0 {
        let t = b;
        b = a % b;
        a = t;
    }
    a
}

fn isqrt(n: i64) -> i64 {
    if n < 0 {
        panic!("isqrt of negative");
    }
    if n == 0 {
        return 0;
    }
    let mut x = (n as f64).sqrt() as i64;
    while x * x > n {
        x -= 1;
    }
    while (x + 1) * (x + 1) <= n {
        x += 1;
    }
    x
}

fn pythagorean_partners(limit: i64) -> Vec<Vec<i64>> {
    let mut partners: Vec<Vec<i64>> = vec![vec![]; (limit + 1) as usize];
    let r_max = isqrt(2 * limit) + 3;
    for r in 2..=r_max {
        let rr = r * r;
        for s in 1..r {
            if ((r - s) & 1) == 0 || gcd(r, s) != 1 {
                continue;
            }
            let a = rr - s * s;
            let b = 2 * r * s;
            let m = if a > b { a } else { b };
            let x = if a > b { b } else { a };
            if m > limit {
                continue;
            }
            let mut km = m;
            while km <= limit {
                partners[km as usize].push((km / m) * x);
                km += m;
            }
        }
    }
    for row in partners.iter_mut() {
        row.sort();
    }
    partners
}

fn is_minimum_square(sides: &[i64; 3], twice_area: i64, square_side: i64) -> bool {
    let ss = sides;
    let m = square_side;
    let m2 = m * m;
    let d_area = twice_area;
    let mut has_equal_candidate = false;

    for i in 0..3 {
        let d = ss[i];
        let e = ss[(i + 1) % 3];
        let f = ss[(i + 2) % 3];
        let den = 2 * d;
        let t_num = d * d + e * e - f * f;
        let d_num = d * den;
        let span_max = *[0i64, d_num, t_num].iter().max().unwrap();
        let span_min = *[0i64, d_num, t_num].iter().min().unwrap();
        let width_num = span_max - span_min;
        let width_cmp = width_num - m * den;
        let height_cmp = d_area - m * d;
        if width_cmp < 0 && height_cmp < 0 {
            return false;
        }
        if width_cmp <= 0 && height_cmp <= 0 && (width_cmp == 0 || height_cmp == 0) {
            has_equal_candidate = true;
        }
    }

    for i in 0..3 {
        let r = ss[i];
        let p = ss[(i + 1) % 3];
        let q = ss[(i + 2) % 3];
        let k_num = p * p + q * q - r * r;
        if k_num <= 0 {
            continue;
        }
        let r_den_part = p * p + q * q - 2 * d_area;
        if r_den_part <= 0 {
            continue;
        }
        let num = k_num * k_num;
        let den = 4 * r_den_part;
        let p2 = p * p;
        let q2 = q * q;
        if num < d_area * den {
            continue;
        }
        if num > p2 * den || num > q2 * den {
            continue;
        }
        if p2 * den > 2 * num || q2 * den > 2 * num {
            continue;
        }
        let target = m2 * den;
        if num < target {
            return false;
        }
        if num == target {
            has_equal_candidate = true;
        }
    }

    has_equal_candidate
}

fn solve(limit: i64) -> i64 {
    let partners = pythagorean_partners(limit);
    let mut seen: HashSet<[i64; 3]> = HashSet::new();
    let mut total: i64 = 0;

    for m in 1..=limit {
        let mm = m * m;
        let mut row: Vec<(i64, i64)> = vec![(0, m)];
        for &x in &partners[m as usize] {
            row.push((x, isqrt(mm + x * x)));
        }
        let row_len = row.len();

        for i in 0..row_len {
            let (x, hx) = row[i];
            for j in i..row_len {
                let (y, hy) = row[j];
                let base = x + y;
                if base == 0 {
                    continue;
                }
                if base > m {
                    break;
                }
                if x * y < m * (m - base) {
                    continue;
                }
                let mut arr = [base, hx, hy];
                arr.sort();
                if !seen.contains(&arr) {
                    seen.insert(arr);
                    total += arr[0] + arr[1] + arr[2];
                }
            }
        }

        for i in 0..row_len {
            let (u, hu) = row[i];
            for j in i..row_len {
                let (v, hv) = row[j];
                let twice_area = mm - u * v;
                if twice_area <= 0 {
                    continue;
                }
                let p = m - u;
                let q = m - v;
                let third2 = p * p + q * q;
                let third = isqrt(third2);
                if third * third != third2 || third == 0 {
                    continue;
                }
                let mut arr = [third, hu, hv];
                arr.sort();
                if seen.contains(&arr) {
                    continue;
                }
                if is_minimum_square(&arr, twice_area, m) {
                    seen.insert(arr);
                    total += arr[0] + arr[1] + arr[2];
                }
            }
        }
    }

    total
}

fn main() {
    assert_eq!(solve(40), 346);
    assert_eq!(solve(400), 76402);
    assert_eq!(solve(2000), 3237036);
    println!("{}", solve(1_000_000));
}